### PR TITLE
#4: Fixed repeated upload race-condition.

### LINKED ISSUE #4: File overwrite error with Exasol 6.2.14
## Situation

When running a matrix test against Exasol 6.2.14 the sync check fails with the following error:
```
2021-03-29 14:05:03.083 [WARNING] Leaving container running since reuse is enabled. Don't forget to stop and remove the container manually using docker rm -f CONTAINER_ID. 
[ERROR] Tests run: 14, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 218.208 s <<< FAILURE! - in com.exasol.bucketfs.SyncAwareBucketIT
[ERROR] testReplaceFile{Path}  Time elapsed: 5.61 s  <<< FAILURE!
java.lang.AssertionError: 
Upload number 2: file B
Expected: "abcdeABCDE\n"
     but: was "0123456789\n"
	at com.exasol.bucketfs.SyncAwareBucketIT.testReplaceFile(SyncAwareBucketIT.java:156)
```

## Acceptance Criteria

* Overwrite test green with Exasol 6.2.14

### LINKED ISSUE #26: Fix code smells in 2.0.0
## Situation

Version 2.0.0 showed 4 code smells in the SONAR report.

## Acceptance Criteria

* Zero code smells on SONAR.

### PR DESCRIPTION
**Pull request by redcatbear**
Closes #4.
Closes #26.

### CODE REVIEW

File: doc/changes/changes_2.0.1.md [hunk 1]
```diff
@@ -0,0 +1,20 @@
+# bucketfs-java 2.0.0, released 2021-04-29
```
Comment (AnastasiiaSergienko) [inline doc/changes/changes_2.0.1.md:1]:
```suggestion
# bucketfs-java 2.0.1, released 2021-04-??
```

File: doc/changes/changes_2.0.1.md [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+Code name: Improved repeated upload reliability
+
+## Summary
+
+In certain situations especially on 6.2.x there was a race condition that made detection of successful reupload to the same path in a bucket unreliable.
+This has now been fixed. This scenario is realistically only happening in integration tests, but there it is likely.
+
+Replaced handwritten dependency list with auto-generated list and linked it in the [README](../../README.md).
```
Comment (AnastasiiaSergienko) [inline doc/changes/changes_2.0.1.md:10]: The link won't work if you copy this to the release announcement.

File: src/main/java/com/exasol/bucketfs/SyncAwareBucket.java [hunk 3]
```diff
@@ -62,7 +66,9 @@ private void delayRepeatedUploadToSamePath(final String extendedPathInBucket) th
                             .toString());
                 }
             }
-        } else {
+        } else
+
```
Comment (AnastasiiaSergienko) [inline src/main/java/com/exasol/bucketfs/SyncAwareBucket.java:70]: Empty line to remove

File: src/main/java/com/exasol/bucketfs/WriteEnabledBucket.java [hunk 4]
```diff
@@ -88,7 +88,7 @@ protected BucketAccessException createUploadInterruptedException(final URI uri)
                 messageBuilder("E-BFSJ-6").message("Interrupted trying to upload {{URI}}.", uri).toString());
     }
 
-    private void recordUploadInHistory(final String pathInBucket) {
+    protected void recordUploadInHistory(final String pathInBucket) {
```
Comment (AnastasiiaSergienko) [inline src/main/java/com/exasol/bucketfs/WriteEnabledBucket.java:91]: Should we add Javadocs for protected methods?

**Review by AnastasiiaSergienko — CHANGES_REQUESTED** [review]
A few small findings

File: doc/changes/changes_2.0.1.md (hunk 2 above)
Comment (redcatbear) [reply doc/changes/changes_2.0.1.md:10]: Removed link. It is not important anyway.

File: src/main/java/com/exasol/bucketfs/SyncAwareBucket.java (hunk 3 above)
Comment (redcatbear) [reply src/main/java/com/exasol/bucketfs/SyncAwareBucket.java:70]: Fixed.

File: src/main/java/com/exasol/bucketfs/WriteEnabledBucket.java (hunk 4 above)
Comment (redcatbear) [reply src/main/java/com/exasol/bucketfs/WriteEnabledBucket.java:91]: Fixed.

**Review by AnastasiiaSergienko — APPROVED** [review]

**Review by AnastasiiaSergienko — CHANGES_REQUESTED** [review]
There are two failed validations of RD:
1)  The file 'changelog.md' doesn't contain the following link. Please add '[2.0.1](changes_2.0.1.md)' to the file.
2) Update the release date to 2021-04-30

**Review by AnastasiiaSergienko — APPROVED** [review]

**Pull request merged**